src/protections/MentionFlood.ts

```diff
@@ -0,0 +1,116 @@
+const PORT_REGEX = "(:[0-9]+)?";
+
+
+export class MentionFlood implements IProtection {
+
+    settings = {};
+
+    private justJoined: { [roomId: string]: { [username: string]: Date; }; } = {};
+    private mention: RegExp;
+
+    constructor() {
+        this.mention = new RegExp(`@${LOCALPART_REGEX}:(${DOMAIN_REGEX}|${IPV4_REGEX}|${IPV6_REGEX})${PORT_REGEX}`, "g");
+    }
+
+    public get name(): string {
+        return 'MentionFlood';
+    }
+
+    public async handleEvent(mjolnir: Mjolnir, roomId: string, event: any): Promise<any> {
+
+        const content = event['content'] || {};
+        const minsBeforeTrusting = config.protections.mentionFlood.minutesBeforeTrusting;
+
+        if (minsBeforeTrusting > 0) {
+            if (!this.justJoined[roomId]) this.justJoined[roomId] = {};
+
+            // When a new member logs in, store the time they joined.  This will be useful
+            // when we need to check if a message was sent within 20 minutes of joining
+            if (event['type'] === 'm.room.member') {
+                if (isTrueJoinEvent(event)) {
+                    const now = new Date();
+                    this.justJoined[roomId][event['state_key']] = now;
+                    LogService.info("MentionFlood", `${event['state_key']} joined ${roomId} at ${now.toDateString()}`);
+                } else if (content['membership'] === 'leave' || content['membership'] === 'ban') {
+                    delete this.justJoined[roomId][event['sender']];
+                }
+
+                return;
+            }
+        }
+
+        if (event['type'] === 'm.room.message') {
+            const message = content['formatted_body'] || content['body'] || null;
+
+            // Check conditions first
+            if (minsBeforeTrusting > 0) {
+                const joinTime = this.justJoined[roomId][event['sender']];
+                if (joinTime) { // Disregard if the user isn't recently joined
+
+                    // Check if they did join recently, was it within the timeframe
+                    const now = new Date();
+                    if (now.valueOf() - joinTime.valueOf() > minsBeforeTrusting * 60 * 1000) {
+                        delete this.justJoined[roomId][event['sender']]; // Remove the user
+                        LogService.info("MentionFlood", `${event['sender']} is no longer considered suspect`);
+                        return;
+                    }
+
+                } else {
+                    // The user isn't in the recently joined users list, no need to keep
+                    // looking
+                    return;
+                }
+            }
+
+
+            // Perform the test
+            const maxMentionsPerMessage = config.protections.mentionFlood.maxMentionsPerMessage;
+            if (message && message.match(this.mention).length > maxMentionsPerMessage) {
```

Comment:
Shouldn't this be `matchAll`?

Reply:
As I do the g flag above, I count the amount of full matches here. The all part should already happen using the g flag. I will double-check this later on my mjolnir.

Reply:
This seems to work now as wanted. The actual issue with it was that I didnt escape the backslashes as required